### PR TITLE
Fix shebang language detection

### LINKED ISSUE #431: Tokei misses some shell files
Tokei isn't counting files (with no extension) as shell whose shebang is `#!/bin/bash`. It's indeed not accounted for in [`get_filetype_from_shebang`](https://github.com/XAMPPRocky/tokei/blob/2c41fdb182b7ba1bcc7c9d1cfc0de569538e7f0d/src/language/language_type.hbs.rs#L458).

Fwiw, the best reference I know of for "find all shell [and other] files" is `ack`, [whose heuristic is](https://beyondgrep.com/documentation/):

Has extensions: .sh .bash .csh .tcsh .ksh .zsh .fish, or first line matches `/^#!.*\b(?:ba|t?c|k|z|fi)?sh\b/`.

### PR DESCRIPTION
**Pull request by solanav**
Closes #431.

Adds the `shebang_paths` property to languages.json.

Also adds `from_file_extension` and `from_shebang_path` to `LanguageType`, making `from_path` more clear.

### CODE REVIEW

**Comment by solanav** [comment]
I ended up using
```rust
match extension {
    Some(extension) => LanguageType::from_file_extension(extension.as_str()),
    None => LanguageType::from_shebang_path(&entry),
}
```
because I had some problems with the `.map().or_else()`, and this way looks clearer.

File: src/language/language_type.hbs.rs [hunk 1]
```diff
@@ -396,29 +418,52 @@ impl LanguageType {
...
+    /// let rust = LanguageType::from_file_extension("rs");
+    ///
+    /// assert_eq!(rust, Some(LanguageType::Rust));
+    /// ```
+    pub fn from_file_extension(extension: &str) -> Option<Self> {
+        match extension {
+            {{~#each languages}}
+                {{~#if this.extensions}}
+                    {{~#each this.extensions}}
+                        "{{~this}}" {{~#unless @last}} | {{~/unless}}
```
Comment (XAMPPRocky) [inline src/language/language_type.hbs.rs]: You can actually write the match OR `|` preceding your pattern and then it is always allowed even when there is only element. This change was introduced after I wrote this original code. This was designed for exactly this case where it's easier for codegen.
```suggestion
                        | "{{~this}}"
```

File: src/language/language_type.hbs.rs [hunk 2]
```diff
@@ -396,29 +418,52 @@ impl LanguageType {
...
+    /// Get language from a shebang. May open and read the file.
+    ///
+    /// ```no_run
+    /// use tokei::LanguageType;
+    ///
+    /// let rust = LanguageType::from_shebang_path("./main.rs");
+    ///
+    /// assert_eq!(rust, Some(LanguageType::Rust));
+    /// ```
+    pub fn from_shebang_path(entry: &Path) -> Option<Self> {
```
Comment (XAMPPRocky) [inline src/language/language_type.hbs.rs]: As with the suggestion above this should be `from_shebang`.

File: src/language/language_type.hbs.rs [hunk 3]
```diff
@@ -484,9 +529,17 @@ pub fn get_filetype_from_shebang(file: &Path) -> Option<&'static str>
...
-        Some("#!/bin/sh") => Some("sh"),
-        Some("#!/bin/csh") => Some("csh"),
-        Some("#!/usr/bin/perl") => Some("pl"),
+        {{#each languages}}
+        {{~#each this.shebang_paths}}
+        Some("{{this}}")
+        {{~#unless @last}}
+        |
+        {{~/unless}}
+        {{~/each}}
```
Comment (XAMPPRocky) [inline src/language/language_type.hbs.rs]: Same suggestion as above.
```suggestion
        {{~#each this.shebang_paths}}
        | Some("{{this}}")
        {{~/each}}
```

File: src/language/language_type.hbs.rs [hunk 4]
```diff
@@ -484,9 +529,14 @@ pub fn get_filetype_from_shebang(file: &Path) -> Option<&'static str>
...
-        Some("#!/bin/sh") => Some("sh"),
-        Some("#!/bin/csh") => Some("csh"),
-        Some("#!/usr/bin/perl") => Some("pl"),
+        {{#each languages}}
+        {{~#each this.shebangs}}
+        | Some("{{this}}")
+        {{~/each}}
+        {{~#if this.shebangs}}
+            => Some("{{~@key}}"),
+        {{~/if}}
```
Comment (XAMPPRocky) [inline src/language/language_type.hbs.rs]:
```suggestion
        {{~#if this.shebangs}}
        {{~#each this.shebangs}}
        | Some("{{this}}")
        {{~/each}}
            => Some("{{~@key}}"),
        {{~/if}}
```

**Comment by XAMPPRocky** [comment]
Looks good! It seems it's now failing the test suite now though.

File: src/language/language_type.hbs.rs [hunk 5]
```diff
@@ -396,29 +418,53 @@ impl LanguageType {
...
+    /// ```no_run
+    /// use tokei::LanguageType;
+    /// use std::path::Path;
+    ///
+    /// let rust = LanguageType::from_shebang(Path::new("./main.rs"));
+    ///
+    /// assert_eq!(rust, Some(LanguageType::Rust));
+    /// ```
+    pub fn from_shebang(entry: &Path) -> Option<Self> {
+        if let Some(filetype) = get_filetype_from_shebang(&entry) {
```
Comment (XAMPPRocky) [inline src/language/language_type.hbs.rs]:
```suggestion
    pub fn from_shebang<P: AsRef<P>>(entry: P) -> Option<Self> {
        if let Some(filetype) = get_filetype_from_shebang(entry.as_ref()) {
```

File: src/language/language_type.hbs.rs [hunk 6]
```diff
@@ -396,29 +418,53 @@ impl LanguageType {
...
+
+    /// Get language from a shebang. May open and read the file.
+    ///
+    /// ```no_run
+    /// use tokei::LanguageType;
+    /// use std::path::Path;
+    ///
+    /// let rust = LanguageType::from_shebang(Path::new("./main.rs"));
+    ///
+    /// assert_eq!(rust, Some(LanguageType::Rust));
```
Comment (XAMPPRocky) [inline src/language/language_type.hbs.rs]:
```suggestion
    /// use std::path::Path;
    ///
    /// let rust = LanguageType::from_shebang("./main.rs");
    ///
    /// assert_eq!(rust, Some(LanguageType::Rust));
```

Comment (solanav) [reply src/language/language_type.hbs.rs]: That won't work as from_shebang expects a Path. Right now it directly takes from_path's argument `entry`. Should I change that?

Comment (solanav) [reply src/language/language_type.hbs.rs]: Just searched what AsRef<...> does, amazing! I'm loving your github workflow btw, congrats on that project.

Comment (XAMPPRocky) [reply src/language/language_type.hbs.rs]: Thanks! I made it as a separate template if you'd like to use it for your own projects. https://github.com/XAMPPRocky/mean-bean-ci-template

**Comment by XAMPPRocky** [comment]
Thank you again for your PR, and congrats on your first contribution! 🎉

**Pull request merged**